sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ObjectsShouldNotBeDisposedMoreThanOnce.cs

```diff
@@ -177,22 +179,45 @@ private ProgramState VisitInvocationExpression(InvocationExpressionSyntax instru
             private ProgramState ProcessStreamDisposingTypes(ProgramState programState, SyntaxNode usingExpression)
             {
                 var newProgramState = programState;
+                var objectCreationExpressions = usingExpression.DescendantNodes()
+                    .OfType<ObjectCreationExpressionSyntax>();
 
-                var arguments = usingExpression.DescendantNodes()
-                    .OfType<ObjectCreationExpressionSyntax>()
-                    .Where(IsStreamDisposingType)
-                    .Select(FirstArgumentOrDefault)
-                    .WhereNotNull();
-
-                foreach (var argument in arguments)
+                foreach (var objectCreationExpression in objectCreationExpressions)
                 {
-                    var streamSymbol = this.semanticModel.GetSymbolInfo(argument.Expression).Symbol;
-                    newProgramState = ProcessDisposableSymbol(newProgramState, argument.Expression, streamSymbol);
+                    var objectType = this.semanticModel.GetSymbolInfo(objectCreationExpression.Type)
+                    .Symbol
+                    .GetSymbolType();
+
+                    if (objectType.DerivesOrImplementsAny(typesDisposingUnderlyingStream) &&
+                        objectCreationExpression.ArgumentList?.Arguments is IEnumerable<ArgumentSyntax> argumentList)
+                    {
+                        // Checks for the 'leaveOpen' argument. It is at position 4 if present for StreamReader constructor, while it is at position 3 for StreamWriter.
+                        // See #2491 for more details
+                        var leaveOpenArgumentIndex = objectType.DerivesOrImplements(KnownType.System_IO_StreamReader) ? 4 : 3;

```

Comment:
I would keep this info in a map where key is the type and values is the index. Like this, if a new type will be added, everything will be in the same place. And you can still use `Dictionary.Keys` instead of `typesDisposingUnderlyingStream` for the call `DerivesOrImplementsAny`